gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordHistoryService.java

```diff
@@ -0,0 +1,135 @@
+        LOGGER.debug("Adding password history entry for user {}", userId);
+        if (passwordSettings == null || !passwordSettings.isPasswordHistoryEnabled()) {
+            return Single.error(new IllegalArgumentException("Password history disabled."));
+        }
+        return repository.findUserHistory(referenceType, referenceId, userId)
+                         .toList()
+                         .flatMap(passwordHistories -> {
+                             if (passwordAlreadyUsed(password, passwordHistories)) {
+                                 return Single.error(() -> PasswordHistoryException.passwordAlreadyInHistory(passwordSettings));
+                             }
+                             int passwordCount = passwordHistories.size();
+                             if (passwordCount >= passwordSettings.getOldPasswords()) {
+                                 passwordHistories.sort(comparing(PasswordHistory::getCreatedAt));
+                                 return repository.delete(passwordHistories.get(0).getId())
+                                                  .andThen(repository.create(getPasswordHistory(referenceType, referenceId, userId, password)));
+                             } else {
+                                 return repository.create(getPasswordHistory(referenceType, referenceId, userId, password));
+                             }
+                         })
+                         .doOnSuccess(a -> auditService.report(AuditBuilder.builder(UserAuditBuilder.class)
+                                                                           .principal(principal)
+                                                                           .type(EventType.PASSWORD_HISTORY_CREATED)))
+                         .doOnError(throwable -> auditService.report(AuditBuilder.builder(UserAuditBuilder.class)
+                                                                                 .principal(principal)
+                                                                                 .type(EventType.PASSWORD_HISTORY_CREATED)
+                                                                                 .throwable(throwable)));
+
+    }
+
+    /**
+     * Checks if a password is already in the history
+     *
+     * @param referenceType    the reference type (DOMAIN, APPLICATION)
+     * @param referenceId      id of the reference
+     * @param userId           id of user for this password
+     * @param password         the password to add
+     * @param passwordSettings domain/application password settings
+     * @return Single containing a {@link Boolean} {@code true} if the password is already present the allotted number of previous passwords.
+     */
+    public Single<Boolean> passwordAlreadyUsed(ReferenceType referenceType, String referenceId, String userId, String password, PasswordSettings passwordSettings) {
```

Comment:
Is this method used somewhere ?

Reply:
It will be for the gateway reset password form.